includes/class-wc-logger.php

```diff
@@ -1,161 +1,230 @@
 <?php
-
 if ( ! defined( 'ABSPATH' ) ) {
 	exit; // Exit if accessed directly
 }
 
 /**
- * Allows log files to be written to for debugging purposes
+ * Provides logging capabilities for debugging purposes.
  *
  * @class          WC_Logger
- * @version        1.6.4
+ * @version        2.0.0
  * @package        WooCommerce/Classes
  * @category       Class
  * @author         WooThemes
  */
 class WC_Logger {
 
 	/**
-	 * Stores open file _handles.
+	 * Stores registered log handlers.
 	 *
 	 * @var array
-	 * @access private
 	 */
-	private $_handles;
+	protected $handlers;
+
+	/**
+	 * Minimum log level this handler will process.
+	 *
+	 * @var int Integer representation of minimum log level to handle.
+	 */
+	protected $threshold;
 
 	/**
 	 * Constructor for the logger.
+	 *
+	 * @param array $handlers Optional. Array of log handlers. If $handlers is not provided,
+	 *     the filter 'woocommerce_register_log_handlers' will be used to define the handlers.
+	 *     If $handlers is provided, the filter will not be applied and the handlers will be
+	 *     used directly.
+	 * @param string $threshold Optional. Define an explicit threshold. Defaults to the global
+	 *     setting 'woocommerce_log_threshold' or 'notice' if the setting is not configured.
 	 */
-	public function __construct() {
-		$this->_handles = array();
+	public function __construct( $handlers = null, $threshold = null ) {
+		if ( null === $handlers ) {
+			$handlers = apply_filters( 'woocommerce_register_log_handlers', array() );
+		}
+
+		if ( null === $threshold ) {
+			$threshold = get_option( 'woocommerce_log_threshold', 'notice' );
+		}
+
+		$this->handlers = $handlers;
+		$this->threshold = WC_Log_Levels::get_level_severity( $threshold );
 	}
 
 	/**
-	 * Destructor.
+	 * Determine whether to handle or ignore log.
+	 *
+	 * @param string $level emergency|alert|critical|error|warning|notice|info|debug
+	 * @return bool True if the log should be handled.
 	 */
-	public function __destruct() {
-		foreach ( $this->_handles as $handle ) {
-			if ( is_resource( $handle ) ) {
-				fclose( $handle );
-			}
-		}
+	public function should_handle( $level ) {
+		return $this->threshold <= WC_Log_Levels::get_level_severity( $level );
 	}
 
 	/**
-	 * Open log file for writing.
+	 * Add a log entry.
+	 *
+	 * @deprecated 2.7.0
 	 *
 	 * @param string $handle
-	 * @param string $mode
-	 * @return bool success
+	 * @param string $message
+	 * @return bool
+	 */
+	public function add( $handle, $message ) {
```

Comment:
Since we have 300 extensions, I wonder if we can take a more laid back approach to this method. i.e. assume this is equivalent to debug level notice. Then extensions can log debug messages as is without using the new log features if they don't need to?